sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java

```diff
@@ -95,6 +95,10 @@ public static int calculateBitSetWidthInBytes(int numFields) {
   }
 
   public static boolean isFixedLength(DataType dt) {
+    if (dt instanceof UserDefinedType) {
+      return isFixedLength(((UserDefinedType) dt).sqlType());
+    }
+
```

Comment:
nit: how about this format?
```
    if (dt instanceof UserDefinedType) {
      return isFixedLength(((UserDefinedType) dt).sqlType());
    } else if (dt instanceof DecimalType) {
      return ((DecimalType) dt).precision() <= Decimal.MAX_LONG_DIGITS();
    } else {
      return mutableFieldTypes.contains(dt);
    }
```

Reply:
I used this to emphasize UDT is handled specially (UDT type vs other types), but it's semantically same and I have no strong voice. Will change.

Reply:
+1 for the merged form. BTW, can we keep the `DecimalType` as the first `if`? I guess `UserDefinedType` can be `else if` statement in the middle.

Reply:
Ordering doesn't matter by syntax, but it affects readability.

Assume we are reading the code line by line when UDT parameter comes in here, we can simply find that it does recursive call with sqlType - then we can just go through with its sqlType without going back. If it is placed in the middle, once we read how UDT is handled like this we have to go back on first to follow how the sqlType is handled.

One alternative approach: maybe it is more verbose, but adding while loop to extract the data type till the data type is UDT would be showing intention cleaner. WDYT?

Reply:
I'm fine with the current code. It's easier to read code when we return earlier for some special cases.